engine-src/Game/LambdaHack/Client/UI/HandleHelperM.hs

```diff
@@ -325,13 +322,15 @@ placesFromState coplace ClientOptions{sexposePlaces} s =
               EM.insertWith addEntries pk (ES.singleton lid, 0, 1, 0) em
             f (PK.PExists pk) em =
               EM.insertWith addEntries pk (ES.singleton lid, 0, 0, 1) em
-        in EM.foldr' f EM.empty lentry
+        in EM.foldr' f EM.empty lentry  -- go through place entrances and depending on place add an entry for whether Entry/Around/Exists - effect being we're counting #s of each type
       insertZeros !em !pk _ = EM.insert pk (ES.empty, 0, 0, 0) em
-      initialPlaces | not sexposePlaces = EM.empty
-                    | otherwise = ofoldlWithKey' coplace insertZeros EM.empty
+      initialPlaces | not sexposePlaces = EM.empty                                -- if not exposing places, don't list them
+                    | otherwise = ofoldlWithKey' coplace insertZeros EM.empty     -- otherwise do
   in EM.unionWith addEntries
        initialPlaces
-       (EM.unionsWith addEntries $ map placesFromLevel $ EM.assocs $ sdungeon s)
+       (EM.unionsWith addEntries $ map placesFromLevel $ EM.assocs $ sdungeon s)  
+        -- flatten the dungeon (which is a map of levelId -> level)
+        -- and accumulate the places for each level with counts
```

Comment:
Not exactly. Rather "gather per-place-kind statistics for each level, then aggregate them over all levels, remembering that the place appeared on the given level (but not how man times)".